### PR TITLE
vimPlugins.markdown-preview-nvim: fix node dependencies

### LINKED ISSUE #137239: vimPlugins.markdown-preview-nvim not working
When running `:MarkdownPreview` I get the following error (silently, I have to run `:mes` to see it)

```
internal/modules/cjs/loader.js:892
  throw err;
  ^
Error: Cannot find module 'tslib'
Require stack:
- /nix/store/5pvg3ncxdr0iqs8qm8j2163khsrl43y5-vimplugin-markdown-preview-nvim-2021-03-10/share/vim-plugins/markdown-preview-nvim/app/lib/app/index.js
- /nix/store/5pvg3ncxdr0iqs8qm8j2163khsrl43y5-vimplugin-markdown-preview-nvim-2021-03-10/share/vim-plugins/markdown-preview-nvim/app/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/nix/store/5pvg3ncxdr0iqs8qm8j2163khsrl43y5-vimplugin-markdown-preview-nvim-2021-03-10/share/vim-plugins/markdown-preview-nvim/app/lib/app/index.js:3:17)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Module.require (internal/modules/cjs/loader.js:961:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/5pvg3ncxdr0iqs8qm8j2163khsrl43y5-vimplugin-markdown-preview-nvim-2021-03-10/share/vim-plugins/markdown-preview-nvim/app/lib/app/index.js',
    '/nix/store/5pvg3ncxdr0iqs8qm8j2163khsrl43y5-vimplugin-markdown-preview-nvim-2021-03-10/share/vim-plugins/markdown-preview-nvim/app/index.js'
  ]
}
```

building my system with 8d8a28b47b7c41aeb4ad01a2bd8b7d26986c3512

@ing @SuperSandro2000 since you're the last person to touch `generated.nix` and I don't know who specifically to @.

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change

Fix #137239 and also add installCheck.
Tested to work with `vim` and `neovim`.

The `package.json` is from https://github.com/iamcco/markdown-preview.nvim/blob/master/app/package.json but I'm not sure how to keep it in sync when auto-update vim plugins. If we fetch the file in vim update script, we may also need to update (all) node packages if dependencies changed, which is quite costly. Also currently no vim plugins requires extra update script besides updating the repo.
Note: the `package.json` need to update only when its dependencies change, since it's only used for `node_modules`. Otherwise, it's fine to be out of sync with upstream.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/misc/vim-plugins/overrides.nix [hunk 1]
```diff
@@ -382,6 +383,30 @@ self: super: {
...
+  in {
+    patches = [
+      (substituteAll {
+        src = ./markdown-preview-nvim/fix-node-paths.patch;
+        node = "${nodejs}/bin/node";
+      })
+    ];
+    postInstall = ''
+      # The node package name is `*-vim` not `*-nvim`.
+      ln -s ${nodeDep}/lib/node_modules/markdown-preview-vim/node_modules $out/app
```
Comment (SuperSandro2000) [inline pkgs/misc/vim-plugins/overrides.nix:400]: Lets move that to some FHS directory and not create /app.

Comment (oxalica) [reply pkgs/misc/vim-plugins/overrides.nix:400]: You should build it first and see what's the current behavior.
All vimPlugins follows vim's package hierarchy, with special meaning directories like `autoload`, `plugin` and etc directly under `$out`. The FHS prefix is removed in https://github.com/NixOS/nixpkgs/commit/56f823dd5c596ef6374f99a22ca63168ff6f6fb9

`app` is a directory from upstream containing a node package. This `ln` creates `$out/app/node_modules`

**Comment by teto** [comment]
thanks for the fix, could you run a nixpkgs-review on this ? I started running one and it broke some things but I dont know if those packages were already broken or if the nodePackages update broke them

File: pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+index 9eebb56..0700333 100644
+--- a/autoload/health/mkdp.vim
++++ b/autoload/health/mkdp.vim
+@@ -9,8 +9,8 @@ function! health#mkdp#check() abort
+     call health#report_info('Pre build: ' . l:mkdp_server_script)
+     call health#report_info('Pre build version: ' . mkdp#util#pre_build_version())
+     call health#report_ok('Using pre build')
+-  elseif executable('node')
+-    call health#report_info('Node version: ' . system('node --version'))
++  elseif executable('@node@')
```
Comment (figsoda) [inline pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch:11]: pretty sure `@node@` always be executable
```suggestion
+  else
```

File: pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+index 5abd807..db1067b 100644
+--- a/autoload/mkdp/nvim/rpc.vim
++++ b/autoload/mkdp/nvim/rpc.vim
+@@ -53,8 +53,8 @@ function! mkdp#nvim#rpc#get_command() abort
+   let l:pre_build = s:root_dir . '/app/bin/markdown-preview-' . mkdp#util#get_platform()
+   if executable(l:pre_build)
+     let l:cmd = [l:pre_build, '--path', s:script]
+-  elseif executable('node')
+-    let l:cmd = ['node', s:root_dir . '/app/index.js', '--path', s:script]
++  elseif executable('@node@')
```
Comment (figsoda) [inline pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch:26]:
```suggestion
+  else
```

File: pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+diff --git a/autoload/mkdp/rpc.vim b/autoload/mkdp/rpc.vim
+index a3361ec..d42f7a6 100644
+--- a/autoload/mkdp/rpc.vim
++++ b/autoload/mkdp/rpc.vim
+@@ -59,9 +59,9 @@ function! mkdp#rpc#start_server() abort
+   let l:mkdp_server_script = s:mkdp_root_dir . '/app/bin/markdown-preview-' . mkdp#util#get_platform()
+   if executable(l:mkdp_server_script)
+     let l:cmd = [l:mkdp_server_script, '--path', s:mkdp_root_dir . '/app/server.js']
+-  elseif executable('node')
++  elseif executable('@node@')
```
Comment (figsoda) [inline pkgs/misc/vim-plugins/markdown-preview-nvim/fix-node-paths.patch:40]:
```suggestion
+  else
```

**Comment by oxalica** [comment]
@teto 
> could you run a nixpkgs-review on this ?

`nixpkgs-review` result on `x86_64-linux`:
```
49 packages updated:
balanceofsatoshis (10.18.1 → 11.0.0) castnow code codium create-cycle-app deltachat-desktop deltachat-desktop EPGStation fast-cli flexoptix-app iosevka joplin lumo mastodon-bot netlify-cli (6.9.15 → 6.9.24) parity-ui pm2 (5.1.1 → 5.1.2) postcss-cli (8.3.1 → 9.0.1) pulp pyright (1.1.170 → 1.1.171) redoc-cli slack slack spacevim teams teleprompter thelounge timedoctor vimPlugins.coc-explorer (0.18.17 → 0.18.18) vimplugin-coc-json vimplugin-coc-metals vimplugin-coc-prettier vimPlugins.coc-pyright (1.1.170 → 1.1.171) vimplugin-coc-stylelint vimplugin-coc-tslint vimplugin-coc-vetur vimPlugins.coc-vimlsp (0.12.4 → 0.12.5) vimplugin-markdown-preview.nvim vscode vscode-extension-MS-python-vscode-pylance vscode-extension-rust-analyzer vscode-extension-vadimcn-vscode-lldb vscode-with-extensions vscodium yaml-language-server zx commitlint (13.1.0 → 13.2.0) _at_hyperspace_slash_cli _at_nerdwallet_slash_shepherd

3 packages failed to build:
hyperspace-cli netlify-cli timedoctor

44 packages built:
balanceofsatoshis castnow commitlint create-cycle-app deltachat-desktop epgstation fast-cli flexoptix-app iosevka joplin lumo mastodon-bot parity-ui pm2 postcss-cli pulp pyright redoc-cli shepherd slack spacevim teams teleprompter thelounge vimPlugins.coc-explorer vimPlugins.coc-json vimPlugins.coc-metals vimPlugins.coc-prettier vimPlugins.coc-pyright vimPlugins.coc-stylelint vimPlugins.coc-tslint vimPlugins.coc-vetur vimPlugins.coc-vimlsp vimPlugins.markdown-preview-nvim vscode vscode-extensions.matklad.rust-analyzer vscode-extensions.ms-python.vscode-pylance vscode-extensions.vadimcn.vscode-lldb vscode-fhs vscode-with-extensions vscodium vscodium-fhs yaml-language-server zx
```

I confirmed the 3 failed packages are already failed on current `master` so they are unrelated.
Here's the log if someone is interested in it: https://gist.github.com/oxalica/7360da26f14cb2b054571d77a35d894b

**Pull request merged**